library/ssl_tls.c

```diff
@@ -3142,6 +3142,38 @@ void mbedtls_ssl_init( mbedtls_ssl_context *ssl )
     memset( ssl, 0, sizeof( mbedtls_ssl_context ) );
 }
 
+static int ssl_conf_check(const mbedtls_ssl_context *ssl)
+{
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+    if( mbedtls_ssl_conf_is_tls13_only( ssl->conf ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 4, ( "The SSL configuration is is tls13 only." ) );
+        return( 0 );
+    }
+#endif
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2)
+    if( mbedtls_ssl_conf_is_tls12_only( ssl->conf ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 4, ( "The SSL configuration is tls12 only." ) );
+        return( 0 );
+    }
+#endif
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2) && defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+    if( mbedtls_ssl_conf_is_hybrid_tls12_tls13( ssl->conf ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "Hybrid TLS 1.2 + TLS 1.3 configurations are not yet supported" ) );
+        return( MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE );
+    }
+#endif
+
+    /* Space for further checks in the future */
```

Comment:
With the current structure of the function, there is no space for further checks because we return immediately for 1.2-only or 1.3-only functions. If we want to take the chance to introduce a function which can also suit future config checks, the structure would be as follows:
```c
static int ssl_conf_version_check( mbedtls_ssl_context const *ssl )
{
   ... /* what is currently ssl_conf_check() */
}

static int ssl_conf_check( mbedtls_ssl_context const *ssl )
{
    int ret;
    ret = ssl_conf_version_check( ssl );
    if( ret != 0 )
        return( ret );

    /* Space for further checks */
 
    return( 0 );
}
```

Reply:
If you feel that's overkill, you can also rename `ssl_conf_check()` to `ssl_conf_version_check()` and remove the `Space for further checks ...` comment.

Reply:
FWIW, one check that would belong here is the presence of timers in the case of DTLS, which we currently check [every time](https://github.com/ARMmbed/mbedtls/blob/f8f88d6b00a25153b0bdb83b52a7e77792fd53d2/library/ssl_msg.c#L1717) we interface with the underlying transport.